File: template_validator.php
#!/usr/bin/env php
<?php
/**
 * Alertmanager Template Validator
 * 
 * This script validates Alertmanager template files for common syntax issues:
 * - Balanced template delimiters {{ and }}
 * - Balanced define/end blocks
 * - Common syntax patterns
 * - Potential HTML escaping issues
 * 
 * Usage: php template_validator.php <template_filename>
 */

if ($argc < 2) {
    echo "Error: Missing template filename.\n";
    echo "Usage: php template_validator.php <template_filename>\n";
    exit(1);
}

$filename = $argv[1];

if (!file_exists($filename)) {
    echo "Error: File '$filename' not found.\n";
    exit(1);
}

echo "Validating template file: $filename\n";
echo "--------------------------------\n";

// Read the template file
$content = file_get_contents($filename);
if ($content === false) {
    echo "Error: Failed to read template file.\n";
    exit(1);
}

$issues = [];
$warnings = [];
$successMessages = [];

// Check for balanced delimiters
$openingDelimiters = substr_count($content, '{{');
$closingDelimiters = substr_count($content, '}}');

if ($openingDelimiters === $closingDelimiters) {
    $successMessages[] = "✓ Template delimiters '{{' and '}}' are balanced ($openingDelimiters pairs)";
} else {
    $issues[] = "✗ Unbalanced template delimiters: {{: $openingDelimiters, }}: $closingDelimiters";
}

// Check for balanced define and end blocks
preg_match_all('/\{\{\s*define\s+["\']([^"\']+)["\']\s*\}\}/', $content, $defineMatches);
$defineCount = count($defineMatches[0]);
$endCount = preg_match_all('/\{\{\s*end\s*\}\}/', $content);

if ($defineCount === $endCount) {
    $successMessages[] = "✓ Define/end blocks are balanced ($defineCount blocks)";
} else {
    $issues[] = "✗ Unbalanced define/end blocks: define: $defineCount, end: $endCount";
    
    // List all defined templates
    echo "\nDefined templates:\n";
    foreach ($defineMatches[1] as $templateName) {
        echo "  - $templateName\n";
    }
}

// Check for potentially missing HTML escaping in variables
preg_match_all('/\{\{\s*\.([a-zA-Z0-9_\.]+)\s*\}\}/', $content, $varMatches);
preg_match_all('/\{\{\s*html\s+\.([a-zA-Z0-9_\.]+)\s*\}\}/', $content, $htmlEscapedMatches);

$unescapedVars = array_diff($varMatches[1], $htmlEscapedMatches[1]);
if (!empty($unescapedVars)) {
    $warnings[] = "⚠ Potential HTML escaping issues. The following variables may need 'html' escaping:";
    foreach (array_unique($unescapedVars) as $var) {
        $warnings[] = "  - .$var";
    }
}

// Check for common syntax issues
$syntaxPatterns = [
    'With blocks' => ['/\{\{\s*with\s+/', '/\{\{\s*end\s*\}\}/'],
    'Range blocks' => ['/\{\{\s*range\s+/', '/\{\{\s*end\s*\}\}/'],
    'If blocks' => ['/\{\{\s*if\s+/', '/\{\{\s*end\s*\}\}/'],
    'Template calls' => ['/\{\{\s*template\s+"[^"]+"/', null],
];

foreach ($syntaxPatterns as $patternName => [$openPattern, $closePattern]) {
    $openCount = preg_match_all($openPattern, $content);
    
    if ($openCount > 0) {
        if ($closePattern !== null) {
            $closeCount = preg_match_all($closePattern, $content);
            if ($openCount === $closeCount) {
                $successMessages[] = "✓ $patternName are properly balanced ($openCount)";
            } else {
                $issues[] = "✗ Unbalanced $patternName: open: $openCount, close: $closeCount";
            }
        } else {
            $successMessages[] = "✓ Found $openCount $patternName";
        }
    }
}

// Check for unusual combinations
if (preg_match('/\{\{\{\s*/', $content)) {
    $warnings[] = "⚠ Found '{{{' pattern - this might be a syntax error or a typo";
}

if (preg_match('/\}\}\}/', $content)) {
    $warnings[] = "⚠ Found '}}}' pattern - this might be a syntax error or a typo";
}

// Output results
echo "\nValidation Results:\n";
echo "==================\n";

if (empty($issues) && empty($warnings)) {
    echo "✅ No issues found! The template appears to be valid.\n";
} else {
    if (!empty($issues)) {
        echo "\nErrors:\n";
        foreach ($issues as $issue) {
            echo "$issue\n";
        }
    }
    
    if (!empty($warnings)) {
        echo "\nWarnings:\n";
        foreach ($warnings as $warning) {
            echo "$warning\n";
        }
    }
}

if (!empty($successMessages)) {
    echo "\nSuccess Checks:\n";
    foreach ($successMessages as $message) {
        echo "$message\n";
    }
}

// Quick content overview
$templateLines = explode("\n", $content);
$lineCount = count($templateLines);
echo "\nTemplate Overview:\n";
echo "Total lines: $lineCount\n";

// Output line count and template structure summary
$definedTemplates = [];
foreach ($defineMatches[1] as $index => $templateName) {
    $definedTemplates[] = $templateName;
}

if (!empty($definedTemplates)) {
    echo "Defined templates: " . implode(", ", $definedTemplates) . "\n";
}

// Exit with status code based on validation
exit(empty($issues) ? 0 : 1);

File: template_validator.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strings"
	"text/template"
)

// TemplateValidator checks Alertmanager templates for syntax issues
type TemplateValidator struct {
	FileName    string
	Content     string
	Errors      []string
	Warnings    []string
	Definitions map[string]bool
}

// NewTemplateValidator creates a new template validator for the given file
func NewTemplateValidator(fileName string) (*TemplateValidator, error) {
	content, err := ioutil.ReadFile(fileName)
	if err != nil {
		return nil, fmt.Errorf("error reading file %s: %v", fileName, err)
	}

	return &TemplateValidator{
		FileName:    fileName,
		Content:     string(content),
		Errors:      []string{},
		Warnings:    []string{},
		Definitions: make(map[string]bool),
	}, nil
}

// Validate performs all validation checks
func (v *TemplateValidator) Validate() bool {
	v.validateTemplateDefinitions()
	v.validateTemplateSyntax()
	v.validateCommonIssues()
	
	return len(v.Errors) == 0
}

// validateTemplateDefinitions checks for properly defined and closed templates
func (v *TemplateValidator) validateTemplateDefinitions() {
	// Find all template definitions
	defineRegex := regexp.MustCompile(`{{-?\s*define\s+"([^"]+)"\s*-?}}`)
	endRegex := regexp.MustCompile(`{{-?\s*end\s*-?}}`)
	
	defines := defineRegex.FindAllStringSubmatch(v.Content, -1)
	endCount := len(endRegex.FindAllString(v.Content, -1))
	
	if len(defines) != endCount {
		v.Errors = append(v.Errors, fmt.Sprintf("Mismatch between 'define' (%d) and 'end' (%d) statements", 
			len(defines), endCount))
	}
	
	for _, match := range defines {
		if len(match) > 1 {
			templateName := match[1]
			if _, exists := v.Definitions[templateName]; exists {
				v.Errors = append(v.Errors, fmt.Sprintf("Duplicate template definition: %s", templateName))
			} else {
				v.Definitions[templateName] = true
			}
		}
	}
}

// validateTemplateSyntax checks if the template can be parsed by Go's template engine
func (v *TemplateValidator) validateTemplateSyntax() {
	_, err := template.New(v.FileName).Parse(v.Content)
	if err != nil {
		v.Errors = append(v.Errors, fmt.Sprintf("Template syntax error: %v", err))
	}
}

// validateCommonIssues checks for common mistakes in Alertmanager templates
func (v *TemplateValidator) validateCommonIssues() {
	// Check for missing variable escaping
	if strings.Contains(v.Content, "{{.") && !strings.Contains(v.Content, "{{.}}") {
		// Look for unescaped variables in places where HTML would be expected
		riskContexts := []string{"title", "text", "summary", "description"}
		
		for _, context := range riskContexts {
			regex := regexp.MustCompile(fmt.Sprintf(`%s[^}]*{{\.([^|]*)}}`, context))
			matches := regex.FindAllString(v.Content, -1)
			
			for _, match := range matches {
				if !strings.Contains(match, "|") && !strings.Contains(match, "reReplaceAll") && !strings.Contains(match, "html") {
					v.Warnings = append(v.Warnings, fmt.Sprintf("Potential unescaped variable in HTML context: %s", match))
				}
			}
		}
	}
	
	// Check for unclosed action blocks
	openActions := strings.Count(v.Content, "{{")
	closeActions := strings.Count(v.Content, "}}")
	if openActions != closeActions {
		v.Errors = append(v.Errors, fmt.Sprintf("Mismatch between opening '{{' (%d) and closing '}}' (%d) delimiters", 
			openActions, closeActions))
	}
	
	// Check for undefined variables or functions
	commonFunctions := []string{"toUpper", "toLower", "title", "reReplaceAll", "join", "printf", "index", "len"}
	definedFunctions := map[string]bool{}
	
	for _, fn := range commonFunctions {
		definedFunctions[fn] = true
	}
	
	funcRegex := regexp.MustCompile(`{{[^}]*\s+(\w+)\s+[^}]*}}`)
	funcMatches := funcRegex.FindAllStringSubmatch(v.Content, -1)
	
	for _, match := range funcMatches {
		if len(match) > 1 {
			function := match[1]
			if !definedFunctions[function] && !strings.HasPrefix(function, ".") && !strings.HasPrefix(function, "define") && function != "if" && function != "else" && function != "end" && function != "range" && function != "template" && function != "with" {
				v.Warnings = append(v.Warnings, fmt.Sprintf("Potential undefined function call: %s", function))
			}
		}
	}
}

// PrintResults prints validation results
func (v *TemplateValidator) PrintResults() {
	if len(v.Errors) == 0 && len(v.Warnings) == 0 {
		fmt.Printf("✅ File %s is valid and free of common issues.\n", v.FileName)
		return
	}
	
	fmt.Printf("Results for %s:\n", v.FileName)
	
	if len(v.Errors) > 0 {
		fmt.Println("\n❌ ERRORS:")
		for i, err := range v.Errors {
			fmt.Printf("%d. %s\n", i+1, err)
		}
	}
	
	if len(v.Warnings) > 0 {
		fmt.Println("\n⚠️  WARNINGS:")
		for i, warning := range v.Warnings {
			fmt.Printf("%d. %s\n", i+1, warning)
		}
	}
	
	fmt.Println()
	if len(v.Definitions) > 0 {
		fmt.Println("Found template definitions:")
		for name := range v.Definitions {
			fmt.Printf("- %s\n", name)
		}
	}
}

func main() {
	// Parse command line arguments
	fileName := flag.String("file", "slack.tmpl", "Template file to validate")
	flag.Parse()
	
	validator, err := NewTemplateValidator(*fileName)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}
	
	isValid := validator.Validate()
	validator.PrintResults()
	
	if !isValid {
		os.Exit(1)
	}
}

